x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/AllocationReducer.java

```diff
@@ -0,0 +1,137 @@
+
+        if (totalRemainingAllocations <= numberOfAllocations) {
+            logger.warn("request to reduce allocations to equal or less the existing number of allocations");
+            throw new IllegalArgumentException("request to reduce allocations to equal or less the existing number of allocations");
+        }
+
+        while (totalRemainingAllocations > numberOfAllocations) {
+            // While we reduce allocations there are 2 concerns:
+            // 1. Remove entire assignments when possible to free the used memory
+            // 2. Preserve balance across zones
+            // The way we achieve this here is simple. Each iteration, we pick the zone with the most allocations
+            // and find its smallest assignment. We then check if we can remove that assignment entirely, or we
+            // reduce its allocations by 1.
+
+            final int allocationsToRemove = totalRemainingAllocations - numberOfAllocations;
+            List<Map.Entry<List<String>, Integer>> allocationsPerZoneInAscendingOrder = allocationsByZone.entrySet()
+                .stream()
+                .sorted(Map.Entry.comparingByValue())
+                .toList();
+            if (allocationsPerZoneInAscendingOrder.isEmpty()) {
+                logger.warn("no allocations remain in any zone");
+                throw new IllegalStateException("no allocations remain in any zone");
+            }
+            final List<String> largestZone = allocationsPerZoneInAscendingOrder.get(allocationsPerZoneInAscendingOrder.size() - 1).getKey();
+            final int largestZoneAllocations = allocationsPerZoneInAscendingOrder.get(allocationsPerZoneInAscendingOrder.size() - 1)
+                .getValue();
+            final int minAllocationsInOtherZones = allocationsPerZoneInAscendingOrder.size() <= 1
+                ? 0
+                : allocationsPerZoneInAscendingOrder.get(0).getValue();
+
+            List<Map.Entry<String, Integer>> largestZoneAssignmentsInAscendingOrder = allocationsByNode.entrySet()
+                .stream()
+                .filter(e -> nodeIdsByZone.get(largestZone).contains(e.getKey()))
+                .sorted(Map.Entry.comparingByValue())
+                .toList();
+
+            if (largestZoneAssignmentsInAscendingOrder.isEmpty()) {
+                logger.warn("no assignments remain in the largest zone");
+                throw new IllegalStateException("no assignments remain in the largest zone");
+            }
+
+            Map.Entry<String, Integer> smallestAssignmentInLargestZone = largestZoneAssignmentsInAscendingOrder.get(0);
+            if (smallestAssignmentInLargestZone.getValue() == 1
+                || (smallestAssignmentInLargestZone.getValue() <= allocationsToRemove
+                    && (minAllocationsInOtherZones == 0
+                        || ((largestZoneAllocations - smallestAssignmentInLargestZone.getValue()) >= minAllocationsInOtherZones + 1)))) {
```

Comment:
```suggestion
            boolean noAllocationsInOtherZones = allocationsPerZoneInAscendingOrder.size() == 1;
            if (smallestAssignmentInLargestZone.getValue() == 1
                || (smallestAssignmentInLargestZone.getValue() <= allocationsToRemove
                    && (noAllocationsInOtherZones
                        || ((largestZoneAllocations - smallestAssignmentInLargestZone.getValue()) > minAllocationsInOtherZones)))) {
```

I found this quite hard to follow could you add a comment please

Reply:
I have refactored the check in a method with commentary. That should make it simpler to understand.